lib/generator.js

```diff
@@ -82,10 +83,12 @@ class Generator {
    * @param {Boolean} [options.debug=false] Enable more specific errors in the console. At the moment it only shows specific errors about filters. Keep in mind that as a result errors about template are less descriptive.
    */
   constructor(templateName, targetDir, { templateParams = {}, entrypoint, noOverwriteGlobs, disabledHooks, output = 'fs', forceWrite = false, install = false, debug = false } = {}) {
+    const invalidOptionsError = invalidOptions(arguments[arguments.length - 1] || []);
+    if (invalidOptionsError) throw new Error(invalidOptionsError);
```

Comment:
we kind of still have it here an error message hidden in the variable which is not the best as while reading this and next line I want to have a clear answer what error is thrown when

best experience would be to have
```
const invalidOptions = getInvalidOptions(GENERATOR_OPTIONS, arguments[arguments.length - 1])
if (getInvalidOptions.length) throw new Error(`These options are not supported by the generator: ${invalidOptions.join(', ')}`);
```

why:
- `GENERATOR_OPTIONS` belongs to a file where the generator constructor is, for maintenance reasons 
- error message reason mentioned at the beginning

then just merge `invalidOptions` with `getInvalidOptions` function that should not be marked as private.

---

Despite my comment above I still have a big concern about how we get `GENERATOR_OPTIONS` up to date, as if I forget to put new options there, no tests will fail. You noticed that in the initial implementation you forgot about `debug`

Reply:
https://github.com/dsheiko/bycontract looks interesting but I haven't tried it, but it talks about jsdoc that we at least always keep up to date

Reply:
Remove `invalidOptions`, really I don't know why both were marked as private, maybe too much time looking to the same piece of code 🙂

> Despite my comment above I still have a big concern about how we get `GENERATOR_OPTIONS` up to date, as if I forget to put new options there, no tests will fail. You noticed that in the initial implementation you forgot about `debug`

That's a good question. It'd be easy if JS provides keyword arguments, so we use them to reference the mandatory ones one and everything else can be passed as an optional object, but it doesn't.

Or we could define the generator options within the class, and dynamically build the constructor by iterating through its element, so it acts as the single source of truth. E.g:

```javascript
class Generator {
  constructor(templateName, targetDir, options) {
    const GENERATOR_OPTIONS = ['debug', 'disabledHooks', 'entrypoint', 'forceWrite', 'install', 'noOverwriteGlobs', 'output', 'templateParams'];
    GENERATOR_OPTIONS.forEach(option => {
      this[option] = options[option];
    });
  }
}

let gen = new Generator('foo', 'bar', { debug: 'a', disabledHooks: 'b', entrypoint: 'c' });
gen.debug            // 'a'
gen.disabledHooks    // 'b'
gen.entrypoint       // 'c'
gen.forceWrite       // undefined
gen.install          // undefined
gen.noOverwriteGlobs // undefined
gen.output           // undefined
gen.templateParams   // undefined
```

Reply:
looks nice but then you lose what is nice currently, that options are destructured in params and we can use defaults there. I'm afraid now adding to your suggesting defaults, would complicate stuff 🤔 
did you try the library I mentioned? sounds like the perfect choice

Reply:
nono I haven't tried it still, I wanted to know if we have to chance to do it without libraries.

Would you like to try the library and incorporate it to this PR or do a separate one for that?

Reply:
nah, let's leave it, I was just curious, to use this type of decorators we would probably have to use babel or something anyway.

Just please move `GENERATOR_OPTIONS` to `generator.js` so we maintain the list where the constructor is, and just extend `getInvalidOptions` with an additional argument.

I think that whatever option we choose here to have "future errors secure implementation" will just add complexity that I'm not sure is worth it. Let's ignore my initial concern and react properly in the future where we see it is really a problem

Reply:
@sebastian-palma we want to release 1.0.0-rc1 and it would be great to have this fix for the CLI in there. Do you agree with my comment and still want to work on it, or did you loose your patience because of me being picky 😅

Reply:
Hey @derberg, just committed and squashed the changes. Sorry for the delay 😁